thmlike order keys and general cleanup (#2163)
latex3/latex2e: base/doc/ltnews44.tex

\texttt{lttemplate} in the kernel in due course.

\subsection{Use the \key{order} key mechanism for theorem-like environments and proofs}

The theorem-like templates have been rewritten to make use of the new
\key{order} key mechanism.  As a result a number of keys got replaced
with new (more standardized) ones. Thus, for \key{title},
\key{number}, \key{note}, and \key{punct} there are now keys for
customization with the names \key{title-decls}, \key{number-decls}, etc., as
well as \key{title-format}, \key{number-format} and so on that expect
an argument.

The default for the \key{order} key in theorem-like environments is
\begin{verbatim}
  order={title,separator-1,number,
         separator-2,note,punct}
\end{verbatim}
To put the number first or the punctuation directly after the
number simply switch the names in the comma list around.

Also note that two distinct separator items are supported (both
defaulting to a space), which means that one can easily alter one or
both spaces individually simply by setting the keys \key{separator-1}
and/or \key{separator-2} on the instance.

The default for \key{order} in proofs is slightly simpler because proofs aren't

> numbered so that the template as no \key{number}, \key{number-decls}

Reviewer: ```suggestion
numbered so that the template has no \key{number}, \key{number-decls}
```

> or \key{number-format} keys and consequentally there can only at most a single

Reviewer: ```suggestion
or \key{number-format} keys and consequently there can be at most a single
```

separator:
\begin{verbatim}
  order={title,separator,note,punct}
\end{verbatim}

This also means proofs can now have a note (which is an extension to \pkg{amsthm})
which could be given as
\begin{verbatim}
  \begin{proof}[note={my note}]
\end{verbatim}
By default the setting of the \key{note-format} key surrounds the note with parentheses.

\subsection{Fix a color leakage in theorem-like environments}

If the \key{body-decls} key was given a color specification then that
color leaked back into the caption of that theorem, because placing
the caption was delayed until a paragraph got started by which time
the body declarations were already in force.\footnote{Color in \TeX{}
is different to, say, font attributes: if you have already placed some
text in a box it still reacts to the current color when the box is
placed into the document. In contrast, font changes would have no

> effect, whatsoever at this point.} This was fixed as part of

Reviewer: ```suggestion
effect whatsoever at this point.} This was fixed as part of
```

> \key{order} key rewrite.

Reviewer: ```suggestion
the \key{order} key rewrite.
```
